### ISSUE TITLE
Galamax

### ISSUE THREAD

**Issue by ghotik** [narzoul/DDrawCompat#408]
I got a weird behavior with this game (full name: "Rapid Glider: the Race to Galamax", it is available on MyAbandonware).
The game seems unable to clear properly some previous frame content and this generates weird effects that may seem wanted in some case, but they are not good at all when playing the game. 
Now DDrawCompat fixes the problem perfectly, both alone and together with DxWnd, but since you were so kind to introduce me to the magic of the d3dumddi interfaces I was wondering if you could suggest me what's missing in the DxWnd partial implementation of your features. Apparently, it seems a z-buffer clearing problem, but I tried to apply all the tricks that I imported from DDrawCompat and none of them seemed able to fix the problem!
In attach two sample screenshots of DxWnd alone and DxWnd+DDrawCompat (though you could hardly believe it, it's about the same screen!) but it would be quite easy to get more and see by yourself.

with DxWnd alone:

![dxwnd](https://github.com/user-attachments/assets/51683cbc-1a5b-4ef0-a9b8-49a491243f1d)

with DxWnd + DDrawCompat

![ddrawc](https://github.com/user-attachments/assets/a12812ec-e3a3-43e7-ae77-74ad4c599044)

**Comment by BEENNath58** [comment]
I have seen a similar type of issue, where the content of the screen is "screeched over", on Nvidia with early Microsoft games (eg Monster Truck Madness, Hellbender).

**Comment by narzoul** [comment]
Well, you can just bisect it like I just did. Most of the commits are available as separate builds since v0.3.2, and this was apparently fixed by a commit between v0.3.2 and v.0.4.0, specifically this one: https://github.com/narzoul/DDrawCompat/commit/32301efd5ce80ee5189b136ceb6b2dd4ec81b4e9

So it's just the good old NVIDIA texture color key bug.

I think I already explained how I "fixed" that somewhere, but quick recap: if none of the texture stages need color keying (track D3DDDITSS_TEXTURECOLORKEYVAL and D3DDDITSS_DISABLETEXTURECOLORKEY states in pfnSetTextureStageState), then force the COLORKEYENABLE render state to 0.

**Comment by BEENNath58** [comment]
> So it's just the good old NVIDIA texture color key bug.

Gho is using Intel for DirectDraw (Intel laptop with Nvidia graphics)

**Comment by narzoul** [comment]
I don't have the same issue with AMD natively, and with Intel HD 770 the game crashes on the loading screen. But I'm not aware of any similar color key bugs on any Intel GPUs, and I don't have an old one to test anymore anyway.

**Comment by ghotik** [comment]
Thanks. A quick browsing of my code showed that I hooked the SetTextureStageState at the ddraw/d3d level (in the various interface versions) but just for tracking purposes and I omitted the pfnSetTextureStageState hook at the d3dumddi level, so there are good chances that I could replicate your fix at either level (maybe at both levels). Please, let the topic open until I finish implementing your suggestion, hopefully I won't need further help, but you never know ...

**Comment by BEENNath58** [comment]
Testing on an Nvidia, the same issue exists: so it most likely is the same problem in both scenarios. Thanks

**Comment by BEENNath58** [comment]
Correct, the 2nd scene is problematic on Nvidia, but not on old Intel.

Here's the game on Intel HD Graphics (Intel 3rd Gen):
![IMG20241204192710](https://github.com/user-attachments/assets/78910ac5-5d37-4319-930e-93f934e04ca7)

**Comment by ghotik** [comment]
A slow progress on my understanding of the problem ....
This game Galamax uses IDirect3DDevice2 that doesn't include interfaces like GetTextureStageState / SetTextureStageState, so any logic based on the texture stages at the application interface level is not possible.
To implement it at the d3dumddi I must emulate the DDC logic by creating a m_app private structure holding at least 8 status values for the possible max number of 8 texture stages, or perhaps an even lighter boolean variable telling if while processing the stages we ever met a condition to force the color key disabling.
Now the ting start to become harder ... the m_app should be a private data of a DeviceState class derived from what and when? I currently don't have derived classes with inheritance from the system classes, I use the function pointer replacement technique but this doesn't allow to add private data, so the whole coding should be rewritten.
Then, once completed this part, the flag should be used in the final rendering phase (in DDC going through DeviceState -> flush -> updateRenderStates -> updateRenderState -> mapRsValue) in the pfnClear callback...
Well,  I'm really impressed by your design and I don't know if I will ever have the courage to implement a replica of all this for DxWnd, but I'd like to be to be comforted to know that at least I got the mechanism clearly enough.

**Comment by elishacloud** [comment]
> This game Galamax uses IDirect3DDevice2 that doesn't include interfaces like GetTextureStageState / SetTextureStageState, so any logic based on the texture stages at the application interface level is not possible.

With IDirect3DDevice2 you can use GetRenderState / SetRenderState instead.  See here for approximations of these:

* [GetRenderState()](https://github.com/elishacloud/dxwrapper/blob/c088a676ced68ee7d0bd1bdd78572ba5ffc837b7/ddraw/IDirect3DDeviceX.cpp#L3009-L3176)
* [SetRenderState()](https://github.com/elishacloud/dxwrapper/blob/c088a676ced68ee7d0bd1bdd78572ba5ffc837b7/ddraw/IDirect3DDeviceX.cpp#L2606-L3007)

As far as the different texture stages, I think it is safe to assume that with IDirect3DDevice2 (and older) only stage '0' is relevant.

**Comment by narzoul** [comment]
> Now the ting start to become harder ... the m_app should be a private data of a DeviceState class derived from what and when?

It is associated with the device handle passed to each D3DDDI_DEVICEFUNCS function as the first parameter. You can create the association when the device handle is returned by pfnCreateDevice and destroy it when pfnDestroyDevice is called. DDrawCompat stores the association in a std::map container (actually it stores a Device class and the DeviceState is a member of that) but the implementation details don't really matter, it's up to you how you want to store it.

> Then, once completed this part, the flag should be used in the final rendering phase (in DDC going through DeviceState -> flush -> updateRenderStates -> updateRenderState -> mapRsValue) in the pfnClear callback...

Again you don't need to copy all of that. Just override D3DDDIRS_COLORKEYENABLE if needed before calling any of the pfnDraw* functions if none of the texture stages have color keying enabled.

Multiple texture stages were introduced by DX6. In older versions, at the runtime level there is only one texture handle specified by the D3DRENDERSTATE_TEXTUREHANDLE render state. Texture color keying is enabled automatically if that texture has a color key set through its DirectDrawSurface interface (e.g. with IDirectDrawSurface::SetColorKey, or assigned during surface creation).

**Comment by ghotik** [comment]
Hurray with a surprise.
While I was trying to implement the suggested correction for the "good old NVIDIA texture color key bug" I accidentally enabled the coding for the other d3dumddi experimental integration in DxWnd and the bug was gone! Apparently, there is something in the hooking of these driver callbacks (there are 16 hooked callbacks, but pfnSetTextureStageState is not among them!) that fixes the bug. Maybe later I could try to exclude them one by one to find which one is the magic one, but for the moment I must admit that I don't know why, but the bug was fixed!
![fixed](https://github.com/user-attachments/assets/675301bf-d364-4528-9adc-624cb17a6b56)

**Comment by ghotik** [comment]
> ... Just override D3DDDIRS_COLORKEYENABLE ...

Damn, where? How?
https://www.funny-emoticons.com/view/61/

**Comment by narzoul** [comment]
With [pfnSetRenderState](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dumddi/nc-d3dumddi-pfnd3dddi_setrenderstate).

**Comment by ghotik** [comment]
Thanks. Because of an error in the Microsoft help pages it was not possible to find that document by a Google search, that callback is referred to as Pfnd3dddiSetrenderstate and I was becoming crazy. The call can effectively turn color key on and off, now I have to refine the coding for the Texture stage analysis and I should be close ...

**Comment by ghotik** [comment]
Finally it works !!
I got now a simplified solution working for Galamax. I made it simpler because I prefer to understand the matter gradually, though I realize that there are limitations in my implementation and that DDC coding is much more elegant and generalized. This is my simplified structure:

1) since the pfnCreateDevice is called only once (and I believe that most of the games are using just one device at a time) I didn't make the association table and added a single global BOOL value bColorKeyDisable initialized to TRUE and reset to TRUE within the pfnClear callback wrapper (the idea is: don't use color key unless proven necessary!)

2) in the pfnSetTextureStageState callback wrapper I added these two lines:
`	if (D3DDDITSS_TEXTURECOLORKEYVAL == data->State) bColorKeyDisable = FALSE;
	if (D3DDDITSS_DISABLETEXTURECOLORKEY == data->State) bColorKeyDisable = TRUE;
`
These statements look curiously different from what I found in DDC wrapper, but there must be a reason for that ... anyway this is how it works. I understand that any analysis of the different texture stages is missing, but since Galamax uses D3D2 I don't expect this should be a problem

3) in the pfnDrawPrimitive wrapper (the only one that Galamax uses) I force the color key status according to the current value of bColorKeyDisable (p.s. in this code, pfnSetRenderState is the ORIGINAL callback pointer) :
`	D3DDDIARG_RENDERSTATE data;
	data.State = D3DDDIRS_COLORKEYENABLE;
	data.Value = bColorKeyDisable ? 0 : 1;	
	pfnSetRenderState(hDevice, (const D3DDDIARG_RENDERSTATE *)&data);
`

So, in conclusion, it works and I owe you a million thanks!
